Default interface members prog guide

BillWagner opened the pull request:
Fixes #14803 
Fixes #12489 
Fixes #16821

Depends on dotnet/samples#1998

Review all occurrences of "interface" in the C# programming guide. Update any that state or imply that interface members cannot contain implementation. They can, starting in C# 8.0

While working on those files, do general cleanup:

- check that all snippets are in a project that compiles.
- Various formatting updates.
- Run acrolinx.

Reviews will be easier with hide whitespace.

BillWagner committed:
explicit interface implementation

This commit updates the explicit interface implementation article to discuss how default interface members must be accessed as explicit interface implementations.

BillWagner committed:
next two articles.

BillWagner committed:
update for default interface members.

BillWagner committed:
Add default interface methods

Default interface methods define new capabilities for inheritance.

BillWagner committed:
Add info on DIM

Interfaces may declare indexers, although this is rare.

BillWagner committed:
add nested types in interfaces

Nested types may be declared in interfaces.

Note: I don't think this is new, but the text implied you could not do that.

BillWagner committed:
update allowed modifers for interfaces

Now that interfaces may provide imlementations, it is necessary for interfaces to have more possible access modifiers.

BillWagner committed:
interfaces overview

This file has numerous changes to support the relaxed rules now that default interface methods are allowed. I did not add new samples because default interface members is a more advanced cenario.

BillWagner committed:
first pass edit on the last files.

BillWagner committed:
run acrolinx

BillWagner committed:
remove extra blank line

BillWagner committed:
remove extra blank line.

pkulikov commented:
@BillWagner do you think the [Methods](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/methods) article should be updated as well? Now it claims:

> Methods are declared in a class or struct by specifying ...

implying that one cannot define a method in an interface.

BillWagner commented:
@pkulikov 

> do you think the Methods article should be updated as well?

Yes. I'll open a new PR for that fix.

BillWagner commented:
closing and reopening for a fresh build

BillWagner commented:
switching reviewers to @Thraka

Thraka commented on docs/csharp/programming-guide/classes-and-structs/polymorphism.md:
```diff
@@ -1,93 +1,108 @@
-  
-- [Knowing When to Use Override and New Keywords](./knowing-when-to-use-override-and-new-keywords.md)  
-  
+> It is recommended that virtual members use `base` to call the base class implementation of that member in their own implementation. Letting the base class behavior occur enables the derived class to concentrate on implementing behavior specific to the derived class. If the base class implementation is not called, it is up to the derived class to make their behavior compatible with the behavior  ...
```
Is there some prescriptive guidance on this we can link to? For example, is there a convention around naming a virtual/abstract member `OnDoSomething` you're indicating a callback for the purpose of being overridden and therefor you don't need to call `base.OnDoSomething`? I've always wondered that. I did a quick search but didn't find anything. Maybe you know..

BillWagner replied:
Andy, I don't know of any prescriptive guidance here.

Thraka reviewed (changes requested):
Few things to fix. And snippets not showing yet.

BillWagner committed:
Apply suggestions from code review

Co-Authored-By: Andy De George <2672110+Thraka@users.noreply.github.com>

BillWagner committed:
fix build warnings

BillWagner commented on docs/csharp/programming-guide/concepts/object-oriented-programming.md:
```diff
@@ -116,22 +74,22 @@ class SampleClass
     {
         // Return the value stored in a field.
-        get { return _sample; }
+        get => _sample;
```
The first sample above shows an auto-implemented property. The text above this paragraph explains that this second sample isn't auto-implemented.

Thraka replied:
Oh I see, I missed the fact that the text above was collapsed and I was reading the wrong intro.

BillWagner committed:
respond to feedback.

BillWagner committed:
fix build warnings

BillWagner commented:
closing and reopening for a new build.

BillWagner commented:
Thanks @Thraka 

I've addressed all the feedback.

BillWagner committed:
Add text for #16821

Since I'd done quite a bit on this article, I added the text necessary for this customer issue as well.

BillWagner committed:
fix typo.

BillWagner committed:
fix typo.
